cmd/history_index.cpp

```diff
@@ -0,0 +1,189 @@
+            }
+            rc = changeset_table->get_next(&mdb_key, &mdb_data);
+        }
+
+        if (rc && rc != MDB_NOTFOUND) { /* MDB_NOTFOUND is not actually an error rather eof */
+            lmdb::err_handler(rc);
+        }
+
+        for (const auto &[key, bm] : bitmaps) {
+            Bytes bitmap_bytes(bm.getSizeInBytes(), '\0');
+            bm.write((char *)bitmap_bytes.data());
+            etl::Entry entry{Bytes((unsigned char *)key.c_str(), key.size()), bitmap_bytes};
+            collector.collect(entry);
+        }
+        bitmaps.clear();
+
+        SILKWORM_LOG(LogInfo) << "Latest Block: " << block_number << std::endl;
+        // Proceed only if we've done something
+        if (collector.size()) {
+            SILKWORM_LOG(LogInfo) << "Started Loading" << std::endl;
+
+            auto target_table{txn->open(index_config, MDB_CREATE)};
+            size_t target_table_rcount{0};
+            lmdb::err_handler(target_table->get_rcount(&target_table_rcount));
+            unsigned int db_flags{target_table_rcount ? 0u : (uint32_t)MDB_APPEND};
+
+            // Eventually load collected items WITH transform (may throw)
+            collector.load(
+                target_table.get(),
+                [](etl::Entry entry, lmdb::Table *history_index_table, unsigned int db_flags) {
+                    auto bm{roaring::Roaring64Map::readSafe((const char *)entry.value.data(), entry.value.size())};
+                    Bytes last_chunk_index(entry.key.size() + 8, '\0');
+                    std::memcpy(&last_chunk_index[0], &entry.key[0], entry.key.size());
+                    boost::endian::store_big_u64(&last_chunk_index[entry.key.size()], UINT64_MAX);
+                    auto previous_bitmap_bytes{history_index_table->get(last_chunk_index)};
+                    if (previous_bitmap_bytes.has_value()) {
+                        bm |= roaring::Roaring64Map::readSafe((const char *)previous_bitmap_bytes->data(),
+                                                              previous_bitmap_bytes->size());
+                        db_flags = 0;
+                    }
+                    while (bm.cardinality() > 0) {
+                        auto current_chunk{db::bitmap::cut_left(bm, db::bitmap::kBitmapChunkLimit)};
+                        // make chunk index
+                        Bytes chunk_index(entry.key.size() + 8, '\0');
+                        std::memcpy(&chunk_index[0], &entry.key[0], entry.key.size());
+                        uint64_t suffix{bm.cardinality() == 0 ? UINT64_MAX : current_chunk.maximum()};
+                        boost::endian::store_big_u64(&chunk_index[entry.key.size()], suffix);
+                        Bytes current_chunk_bytes(current_chunk.getSizeInBytes(), '\0');
+                        current_chunk.write((char *)&current_chunk_bytes[0]);
+                        history_index_table->put(chunk_index, current_chunk_bytes, db_flags);
+                    }
+                },
```

Comment:
If I am reading this correctly the transform lambda function is pretty redundant.
Specifically `chunk_index` is the copy of above created `last_chunk_index` (i.e. they both have data from `entry`) with the only difference that in every loop the last 8 bytes of chunk_index are either `UINT32_MAX` or `current_chunk.maximum()`.
Maybe refactoring a bit this lambda can save a lot of `memcpy`